modules/v2x/v2x_proxy/app/v2x_proxy.cc

```diff
@@ -13,147 +13,201 @@
+
+    auto now_us = ::apollo::cyber::Time::MonoTime().ToMicrosecond();
+    if (last_os_light_ == nullptr ||
+        2000LL * 1000 * 1000 < now_us - ts_last_os_light_) {
+      AWARN << "V2X Traffic Light is too old!";
+      last_os_light_ = nullptr;
+    } else {
+      ADEBUG << "V2X Traffic Light is on time.";
+      last_os_light = std::make_shared<OSLight>();
+      last_os_light->CopyFrom(*last_os_light_);
     }
   }
-  if (signals.size() == 1) {
-    auto single = msg->mutable_single_traffic_light(0);
-    single->set_id(signals[0]->id().id());
-    return true;
-  }
-  auto color = msg->single_traffic_light(0).color();
-  msg->clear_single_traffic_light();
-
-  for (auto i = signals.begin(); i != signals.end(); i++) {
-    auto single = msg->add_single_traffic_light();
-    single->set_id((*i)->id().id());
-    single->set_color(color);
+  // proc planning message
+  bool res_proc_planning_msg = internal_->ProcPlanningMessage(
+      planning_adc.get(), last_os_light.get(), &res_light);
+  if (!res_proc_planning_msg) {
+    return;
   }
-  return true;
+  os_interface_->SendV2xTl4hmiToOs(res_light);
 }
 
-void V2xProxy::OnX2vTrafficLightTimer() {
-  x2v_trafficlight_->Clear();
-  obu_interface_->GetV2xTrafficLightFromObu(x2v_trafficlight_);
-  if (!x2v_trafficlight_->has_current_lane_trafficlight()) {
-    AERROR << "Error:v2x trafficlight ignore, no traffic light contained.";
-    return;
-  }
-  auto current_traff = x2v_trafficlight_->mutable_current_lane_trafficlight();
-  if (current_traff->single_traffic_light().empty()) {
-    AERROR << "Error:v2x trafficlight ignore, no traffic light contained.";
-    return;
+void V2xProxy::RecvTrafficlight() {
+  // GET tl from obu
+  std::shared_ptr<OBULight> x2v_traffic_light = nullptr;
+  obu_interface_->GetV2xTrafficLightFromObu(&x2v_traffic_light);
+  os_interface_->SendV2xObuTrafficLightToOs(x2v_traffic_light);
+  auto os_light = std::make_shared<OSLight>();
+  std::string junction_id = "";
+  {
+    std::lock_guard<std::mutex> g(lock_hdmap_junction_id_);
+    junction_id = hdmap_junction_id_;
   }
-  ADEBUG << x2v_trafficlight_->DebugString();
-  if (!TrafficLightProc(current_traff)) {
+  bool res_success_ProcTrafficlight = internal_->ProcTrafficlight(
+      hdmap_, x2v_traffic_light.get(), junction_id, u_turn_,
+      FLAGS_traffic_light_distance, FLAGS_check_time, &os_light);
+  if (!res_success_ProcTrafficlight) {
     return;
   }
-  os_interface_->SendV2xTrafficLightToOs(x2v_trafficlight_);
+  utils::UniqueOslight(os_light.get());
+  os_interface_->SendV2xTrafficLightToOs(os_light);
+  // save for hmi
+  std::lock_guard<std::mutex> lock(lock_last_os_light_);
+  ts_last_os_light_ = ::apollo::cyber::Time::MonoTime().ToMicrosecond();
+  last_os_light_ = os_light;
+}
+
+double cal_distance(const ::apollo::common::PointENU &p1,
+                    const ::apollo::common::PointENU &p2) {
+  double x = p1.x() - p2.x();
+  double y = p1.y() - p2.y();
```

Comment:
Isn't there a distance member function for PointENU class?

Reply:
No. PointENU is a message of proto